go/cmd/dolt/commands/blame.go

```diff
@@ -0,0 +1,450 @@
+	}
+	newTable, err := maybeTableFromCommit(ctx, new, tableName)
+	if err != nil {
+		return false, fmt.Errorf("error getting table from new commit %s: %v", newHash, err)
+	}
+
+	// if the table is in the old commit but not the new one...something's wrong. bail!
+	if oldTable != nil && newTable == nil {
+		return false, fmt.Errorf("expected to find table with name %v in new commit %s, but didn't", tableName, newHash)
+	}
+	// if the table is in the new commit but not the old one, it must be new; return true
+	if newTable != nil && oldTable == nil {
+		return true, nil
+	}
+
+	oldRow, err := maybeRowFromTable(ctx, oldTable, rowPK)
+	if err != nil {
+		return false, fmt.Errorf("error getting row from %s in old commit %s: %v", tableName, oldHash, err)
+	}
+	newRow, err := maybeRowFromTable(ctx, newTable, rowPK)
+	if err != nil {
+		return false, fmt.Errorf("error getting row from %s in new commit %s: %v", tableName, newHash, err)
+	}
+
+	// if the row is in the old table but not the new one...something's wrong. bail!
+	if oldRow != nil && newRow == nil {
+		return false, fmt.Errorf("expected to find row with PK %v in table %s in new commit %s, but didn't", rowPK, tableName, newHash)
+	}
+	// if the row is in the new table but not the old one, it must be new; return true
+	if newRow != nil && oldRow == nil {
+		return true, nil
+	}
+
+	oldSchema, err := oldTable.GetSchema(ctx)
+	if err != nil {
+		return false, fmt.Errorf("error getting schema from %s in old commit %s: %v", tableName, oldHash, err)
+	}
+
+	return !row.AreEqual(*oldRow, *newRow, oldSchema), nil
+}
+
+func blameGraphFromRows(ctx context.Context, rows types.Map) (*blameGraph, error) {
+	graph := make(blameGraph)
+	err := rows.IterAll(ctx, func(key, _ types.Value) error {
+		hash, err := key.Hash(types.Format_7_18)
+		if err != nil {
+			return err
+		}
+		graph[hash] = blameInfo{Key: key}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &graph, nil
+}
+
+// UnblamedNodeKeys returns a slice of primary keys which do not have blame information
+func (bg *blameGraph) UnblamedNodeKeys() []types.Value {
+	keys := []types.Value{}
+	for _, v := range *bg {
+		if v.CommitHash == "" {
+			keys = append(keys, v.Key)
+		}
+	}
+	return keys
+}
+
+// AllNodesBlamed returns true if there are no unblamed nodes in the blame graph
+func (bg *blameGraph) AllNodesBlamed() bool {
+	return len(bg.UnblamedNodeKeys()) == 0
+}
+
+// BlameNode updates the blame graph to contain blame information from the given commit
+// for the row identified by the given primary key
+func (bg *blameGraph) BlameNode(rowPK types.Value, c *doltdb.Commit) error {
+	commitHash, err := c.HashOf()
+	if err != nil {
+		return fmt.Errorf("error getting commit hash: %v", err)
+	}
+
+	meta, err := c.GetCommitMeta()
+	if err != nil {
+		return fmt.Errorf("error getting metadata for commit %s: %v", commitHash.String(), err)
+	}
+
+	pkHash, err := rowPK.Hash(types.Format_7_18)
```

Comment:
Can we thread the right nbf from the database here?

Reply:
Dude I had no idea what I was doing here, happy to talk about it.

Reply:
You can take it from the table or something:

`go/libraries/doltcore/doltdb/table.go`:

```go
func (t *Table) Format() *types.NomsBinFormat {
        return t.vrw.Format()
}
```

It's just a sentinel value that will effect how values serialize themselves at the noms layer.

Reply:
Perfect, thank you